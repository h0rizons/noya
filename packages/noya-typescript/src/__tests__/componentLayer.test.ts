import {
  createBaseFileSystem,
  createTypescriptEnvironment,
  getComponentLayer,
  setAttributeStringValue,
  setFunctionName,
  TypescriptEnvironment,
} from 'noya-typescript';
import ts, { SourceFile } from 'typescript';

let env: TypescriptEnvironment;

beforeAll(async () => {
  const baseFileSystem = await createBaseFileSystem();
  env = createTypescriptEnvironment(baseFileSystem);
});

const filename = 'test.tsx';

function printSourceFile(sourceFile: SourceFile) {
  const printer = ts.createPrinter();

  return printer.printNode(ts.EmitHint.Unspecified, sourceFile, sourceFile);
}

it('gets component layer', () => {
  env.environment.createFile(
    filename,
    `export default function Foo() {
    return <View></View>
  }`,
  );

  const sourceFile = env.environment.getSourceFile(filename)!;

  const componentLayer = getComponentLayer(sourceFile, 'abc');

  expect(componentLayer).toMatchSnapshot();
});

it('gets component layer with attributes', () => {
  env.environment.createFile(
    filename,
    `export default function Foo() {
    return <View name={"test"}></View>
  }`,
  );

  const sourceFile = env.environment.getSourceFile(filename)!;

  const componentLayer = getComponentLayer(sourceFile, 'abc');

  expect(componentLayer).toMatchSnapshot();
});

it('set element name', () => {
  env.environment.createFile(
    filename,
    `export default function Foo() {
    return <View name={"test"}></View>
  }`,
  );

  const sourceFile = env.environment.getSourceFile(filename)!;

  const componentLayer = getComponentLayer(sourceFile, 'abc');
  const attribute = componentLayer!.element.attributes['name'];

  if (attribute.type !== 'stringLiteral') throw new Error('Bad attribute');

  const result = setAttributeStringValue(
    sourceFile,
    attribute.indexPath,
    'hello',
  );

  expect(printSourceFile(result)).toMatchSnapshot();
});

it('visits', () => {
  env.environment.createFile(
    filename,
    `export default function Foo() {
    return <View></View>
  }`,
  );

  const sourceFile = env.environment.getSourceFile(filename)!;

  const updated = setFunctionName(sourceFile, 'Bar');

  const printer = ts.createPrinter();

  const result = printer.printNode(
    ts.EmitHint.Unspecified,
    updated,
    updated as SourceFile,
  );

  expect(result).toMatchSnapshot();
});
