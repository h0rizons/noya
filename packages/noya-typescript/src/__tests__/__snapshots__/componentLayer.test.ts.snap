// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gets component layer 1`] = `
Object {
  "element": Object {
    "attributes": Object {},
    "children": Array [],
    "id": "abc#0",
    "tagName": "View",
  },
  "name": "Foo",
}
`;

exports[`gets component layer with attributes 1`] = `
Object {
  "element": Object {
    "attributes": Object {
      "name": Object {
        "indexPath": Array [
          0,
          3,
          0,
          0,
          0,
          1,
          0,
          1,
        ],
        "type": "stringLiteral",
        "value": "test",
      },
    },
    "children": Array [],
    "id": "abc#0",
    "tagName": "View",
  },
  "name": "Foo",
}
`;

exports[`set element name 1`] = `
"export default function Foo() {
    return <View name=\\"hello\\"></View>;
}
"
`;

exports[`visits 1`] = `
"export default function Bar() {
    return <View></View>;
}
"
`;
